tempesta_fw/sched.c

```diff
@@ -71,6 +70,166 @@ tfw_sched_get_srv_conn(TfwMsg *msg)
+
+	if (st_conn->srv_conn) {
+		/* Failed to sched from the same server. */
+		TfwServer *srv = (TfwServer *)st_conn->srv_conn->peer;
+		char addr_str[TFW_ADDR_STR_BUF_SIZE] = { 0 };
+
+		tfw_addr_ntop(&srv->addr, addr_str, sizeof(addr_str));
+
+		if (!(srv->sg->flags & TFW_SRV_STICKY_FAILOVER)) {
+			TFW_ERR("sched %s: Unable to schedule new request in "
+				"session to server %s in group %s\n",
+				srv->sg->sched->name, addr_str, srv->sg->name);
+			return NULL;
+		}
+		else {
+			TFW_WARN("sched %s: Unable to schedule new request in "
+				 "session to server %s in group %s,"
+				 " fallback to a new server\n",
+				 srv->sg->sched->name, addr_str, srv->sg->name);
+		}
+	}
+
+	write_lock(&st_conn->lock);
+	/*
+	 * Connection and server may return back online while we were trying
+	 * for a lock.
+	 */
+	if ((srv_conn = __try_conn(msg, st_conn)))
+		goto done;
+
+	if (st_conn->main_sg)
+		srv_conn = tfw_sched_get_sg_srv_conn(msg, st_conn->main_sg,
+						     st_conn->backup_sg);
+	else
+		srv_conn = __get_srv_conn(msg);
+
+	/*
+	 * Save connection into session only if used server group is configured
+	 * to have sticky connections.
+	 */
+	if (srv_conn
+	    && (((TfwServer *)srv_conn->peer)->sg->flags & TFW_SRV_STICKY)) {
+		st_conn->srv_conn = srv_conn;
+	}
+
+done:
+	write_unlock(&st_conn->lock);
+
+	return srv_conn;
+}
+
+/*
+ * Find an outgoing server connection for an HTTP message.
+ *
+ * This function is always called in SoftIRQ context.
+ */
+TfwSrvConn *
+tfw_sched_get_srv_conn(TfwMsg *msg)
+{
+	TfwHttpReq *req = (TfwHttpReq *)msg;
+	TfwHttpSess *sess = req->sess;
+
+	/* Sticky cookies are disabled or client doesn't support cookies. */
+	if (!sess)
+		return __get_srv_conn(msg);
+
+	return __get_sticky_srv_conn(msg, sess);
+}
+
+/**
+ * Find connection for a message @msg in @main_sg or @backup_sg server groups.
+ *
+ * If client support cookies the function is called under session's write lock,
+ * and updates to session informations are possible.
+ */
+TfwSrvConn *
+tfw_sched_get_sg_srv_conn(TfwMsg *msg, TfwSrvGroup *main_sg,
+			  TfwSrvGroup *backup_sg)
+{
+	TfwHttpReq *req = (TfwHttpReq *)msg;
+	TfwSrvConn *srv_conn;
+
+	BUG_ON(!main_sg);
+	TFW_DBG2("sched: use server group: '%s'\n", sg->name);
+
+	if (req->sess && (main_sg->flags & TFW_SRV_STICKY)) {
+		TfwStickyConn *st_conn = &req->sess->st_conn;
+
+		/*
+		 * @st_conn->lock is already acquired for writing, if called
+		 * from @__get_sticky_srv_conn().
+		 */
+		st_conn->main_sg = main_sg;
+		st_conn->backup_sg = backup_sg;
+	}
+
+	srv_conn = main_sg->sched->sched_sg_conn(msg, main_sg);
+
+	if (unlikely(!srv_conn && backup_sg)) {
+		TFW_DBG("sched: the main group is offline, use backup: '%s'\n",
+			sg->name);
+		srv_conn = backup_sg->sched->sched_sg_conn(msg, backup_sg);
+	}
+
+	if (unlikely(!srv_conn))
+		TFW_DBG2("sched: Unable to select server from group '%s'\n",
+			 sg->name);
+
+	return srv_conn;
+}
+EXPORT_SYMBOL(tfw_sched_get_sg_srv_conn);
```

Comment:
Please define the function before the first usage of it, i.e. before `__get_sticky_srv_conn()`.